.github/workflows/pr-lint.yml

```diff
@@ -48,7 +48,9 @@ jobs:
       - name: Check PR labels
         env:
           LABEL_NAMES: ${{ toJson(github.event.pull_request.labels.*.name) }}
-        if: contains(env.LABEL_NAMES, 'under review') == contains(env.LABEL_NAMES, 'in progress')
+        if: "!((contains(env.LABEL_NAMES, 'pending review') && !contains(env.LABEL_NAMES, 'in progress') && !contains(env.LABEL_NAMES, 'blocked'))
```

Comment:
maybe you could use '!=' as 'xor' ?
```
if: ! (contains(env.LABEL_NAMES, 'pending review') != contains(env.LABEL_NAMES, 'in progress') != 
   contains(env.LABEL_NAMES, 'blocked'))
   ...
   exit 1
```

Reply:
`xor` would treat `true`, `true`, `true` the same as `true`, `false`, `false`. that said im not sure how an expression with multiple `==` or `!=` execute (they always confuse me so i just pretend they never existed xD) so you are using `xor` here as an analogy or they actually have the very same effect.
solutions i could think of:
- collect these results in a list and count how many `true`s we have.
- add up the `contains` result and make sure they are equal to `1`, booleans should support loose type coercion (`false` -> 0, `true` -> 1): https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#operators

Reply:
Yes 'xor' would not work. Looks like there is no '+' operator either.
I guess we need to set env vars to sum them (like it was done in one of the recent commits)

Reply:
oh. indeed `+` isn't listed in the op list.

maybe we can just avoid github actions' `if` clause and write the complete logic inside the `run:` in good old bash instead.

Reply:
@dimxy @mariocynicys 

I suggest to move to bash if we decide to make lint logic more complex.
bash is more flexible, but less readable with more code lines required.

If we all agree with current labels logic: 
> Allow any number of labels, but only one of the following labels can be present at a time: `['status: pending review', 'status: in progress', 'status: blocked']`. These labels cannot be set together.

then I suggest to leave GH actions expression.

Reply:
> bash is more flexible, but less readable with more code lines required.

i would argue that the current check is harder to read than if it was written in bash.

Reply:
> > bash is more flexible, but less readable with more code lines required.
> 
> i would argue that the current check is harder to read than if it was written in bash.

it would look like this
```bash
- name: Check PR labels
  env:
    LABEL_NAMES: ${{ toJson(github.event.pull_request.labels.*.name) }}
  run: |
    # Convert LABEL_NAMES (JSON) to a string
    LABEL_NAMES_STR=$(echo "${LABEL_NAMES}" | jq -r '.[]')

    # Define allowed labels
    ALLOWED_LABELS=("status: pending review" "status: in progress" "status: blocked")
    MATCHED_COUNT=0

    # Check for matches in LABEL_NAMES
    for label in "${ALLOWED_LABELS[@]}"; do
      if echo "$LABEL_NAMES_STR" | grep -Fxq "$label"; then
        ((MATCHED_COUNT++))
      fi
    done

    # Validate exactly one match
    if [[ "$MATCHED_COUNT" -ne 1 ]]; then
      echo "PR must have exactly one of these labels: ['status: pending review', 'status: in progress', 'status: blocked']."
      exit 1
    fi
```

this is too much, github actions allow to do the same in a more light weight way right in if block without waisting execution time on run

if we need to just be sure that labels have only one of 3 necessary labels this is simpler
```bash
      - name: Check PR labels
        env:
          LABEL_NAMES: ${{ toJson(github.event.pull_request.labels.*.name) }}
        if: "!((contains(env.LABEL_NAMES, 'pending review') && !contains(env.LABEL_NAMES, 'in progress') && !contains(env.LABEL_NAMES, 'blocked'))
          || (!contains(env.LABEL_NAMES, 'pending review') && contains(env.LABEL_NAMES, 'in progress') && !contains(env.LABEL_NAMES, 'blocked'))
          || (!contains(env.LABEL_NAMES, 'pending review') && !contains(env.LABEL_NAMES, 'in progress') && contains(env.LABEL_NAMES, 'blocked')))"
        run: |
          echo "PR must have "exactly one" of these labels: ['status: pending review', 'status: in progress', 'status: blocked']."
          exit 1
```

Reply:
Maybe we can do 2 separate checks:
we have only one PR label starting with "status:" (if not then error 'duplicate status')
PR labels contain any from the list of allowed labels.

Reply:
> we have only one PR label starting with "status:" (if not then error 'duplicate status')

I dont see problem of having `status: pending review` and `status: pending contract deployment` or `status: pending infra` etc together.
Only one status i doubt is `status: invalid` I would move it to `PR must have "exactly one" of these labels` list

> PR labels contain any from the list of allowed labels.

I dont understand what do you mean, above you suggested to have only one status label

Reply:
> I dont understand what do you mean, above you suggested to have only one status label

I mean if we have already ensured that the PR has only single "status:" label,
it is easy to check that this label is one from allowed, like: `contains(env.LABEL_NAMES, 'pending review') || contains(env.LABEL_NAMES, 'in progress') || contains(env.LABEL_NAMES, 'blocked')`

Reply:
> > I dont understand what do you mean, above you suggested to have only one status label
> 
> I mean if we have already ensured that the PR has only single "status:" label, it is easy to check that this label is one from allowed, like: `contains(env.LABEL_NAMES, 'pending review') || contains(env.LABEL_NAMES, 'in progress') || contains(env.LABEL_NAMES, 'blocked')`

these also will lead to move the whole logic to run block using bash, while we can do the same right in one `if` block using GH actions expressions.

```yaml
      - name: Determine label validity
        id: label_check
        env:
          LABEL_NAMES: ${{ toJson(github.event.pull_request.labels.*.name) }}
        run: |
          status_count=$(echo "$LABEL_NAMES" | jq '[.[] | select(startswith("status:"))] | length')

          allowed_status_count=$(echo "$LABEL_NAMES" | jq '[.[] | select(. == "status: pending review" or . == "status: in progress" or . == "status: blocked")] | length')

          if [ "$status_count" -eq 1 ] && [ "$allowed_status_count" -eq 1 ]; then
            # Valid scenario
            echo "valid=true" >> $GITHUB_OUTPUT
          else
            # Invalid scenario
            echo "valid=false" >> $GITHUB_OUTPUT
          fi

      - name: Check PR labels
        if: steps.label_check.outputs.valid == 'false'
        run: |
          echo "PR must have exactly one of these labels: ['status: pending review', 'status: in progress', 'status: blocked']."
          exit 1
```